From 29caa65189f2112a621db5d126602d2aeca20876 Mon Sep 17 00:00:00 2001
From: Gu Chao <gux.chao@intel.com>
Date: Fri, 5 Jun 2020 18:01:42 +0800
Subject: [PATCH] DRV_DPM_SW-298: FLM DPM Klocwork issue cleanup

- Fix assignment in condition issue due to copy_from_user() inside if statement
- Fix various code is unreachable issues
---
 drivers/net/datapath/dpm/datapath_api.c            | 27 ++++++++++++--------
 .../datapath/dpm/gswip30/datapath_lookup_proc.c    | 20 +++------------
 .../datapath/dpm/gswip31/datapath_lookup_proc.c    | 29 ++++++++--------------
 drivers/net/datapath/umt/umt.c                     |  5 ----
 4 files changed, 31 insertions(+), 50 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath_api.c b/drivers/net/datapath/dpm/datapath_api.c
index 8d9c7c5a04cf..7b5e172a4a47 100644
--- a/drivers/net/datapath/dpm/datapath_api.c
+++ b/drivers/net/datapath/dpm/datapath_api.c
@@ -269,18 +269,25 @@ struct pmac_port_info *get_port_info_via_dp_port(int inst, int dp_port)
 
 int8_t parser_size(int8_t v)
 {
-	if (v == DP_PARSER_F_DISABLE)
-		return 0;
-
-	if (v == DP_PARSER_F_HDR_ENABLE)
-		return PASAR_OFFSETS_NUM;
+	int ret;
 
-	if (v == DP_PARSER_F_HDR_OFFSETS_ENABLE)
-		return PASAR_OFFSETS_NUM + PASAR_FLAGS_NUM;
+	switch (v) {
+	case DP_PARSER_F_DISABLE:
+		ret = 0;
+		break;
+	case DP_PARSER_F_HDR_ENABLE:
+		ret = PASAR_OFFSETS_NUM;
+		break;
+	case DP_PARSER_F_HDR_OFFSETS_ENABLE:
+		ret = PASAR_OFFSETS_NUM + PASAR_FLAGS_NUM;
+		break;
+	default:
+		pr_err("Wrong parser setting: %d\n", v);
+		ret = -1;
+		break;
+	}
 
-	pr_err("Wrong parser setting: %d\n", v);
-	/*error */
-	return -1;
+	return ret;
 }
 
 /*Only for SOC side, not for peripheral device side */
diff --git a/drivers/net/datapath/dpm/gswip30/datapath_lookup_proc.c b/drivers/net/datapath/dpm/gswip30/datapath_lookup_proc.c
index 1068bf3be188..22b41f578776 100644
--- a/drivers/net/datapath/dpm/gswip30/datapath_lookup_proc.c
+++ b/drivers/net/datapath/dpm/gswip30/datapath_lookup_proc.c
@@ -386,10 +386,10 @@ int lookup_dump30(struct seq_file *s, int pos)
 	return pos;
 }
 
-ssize_t proc_get_qid_via_index30(struct file *file, const char *buf,
+ssize_t proc_get_qid_via_index30(struct file *file, const char __user *buf,
 				 size_t count, loff_t *ppos)
 {
-	int err = 0, len = 0;
+	size_t len;
 	char data[100];
 	unsigned int lookup_index;
 	unsigned int qid = 0;
@@ -400,20 +400,8 @@ ssize_t proc_get_qid_via_index30(struct file *file, const char *buf,
 		return -EPERM;
 	len = (count >= sizeof(data)) ? (sizeof(data) - 1) : count;
 	DP_DEBUG(DP_DBG_FLAG_LOOKUP, "len=%d\n", len);
-
-	if (len <= 0) {
-		err = -EFAULT;
-		pr_err("Wrong len value (%d)\n", len);
-		return count;
-	}
-
-	if (copy_from_user(data, buf, len)) {
-		err = -EFAULT;
-		pr_err("copy_from_user fail");
-		return count;
-	}
-
-	data[len - 1] = 0;	/* Make string */
+	len -= copy_from_user(data, buf, len);
+	data[len] = 0; /* Make string */
 	num = dp_split_buffer(data, param_list, ARRAY_SIZE(param_list));
 
 	if (num <= 1)
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_lookup_proc.c b/drivers/net/datapath/dpm/gswip31/datapath_lookup_proc.c
index 3160ba9d2109..c75c60f5ff0a 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_lookup_proc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_lookup_proc.c
@@ -141,12 +141,14 @@ static void lookup_table_recursive(int k, int tmp_index, int set_flag, int qid);
 
 /* The main function that prints all combinations of size r*/
 /* in arr[] of size n. This function mainly uses combine_util()*/
-static void c_not_care_walkthrought(int *arr, int n, int r)
+static int c_not_care_walkthrought(int *arr, int n, int r)
 {
 	/* A temporary array data[] to store all combination one by one */
 
 	/* Print all combination using temprary array 'data[]' */
 	combine_util(arr, c_tmp_data, 0, n - 1, 0, r);
+	/* return lookup_match_num here to avoid KW warning */
+	return lookup_match_num;
 }
 
 /* arr[]  ---> Input Array
@@ -327,6 +329,7 @@ int find_pattern(int port_id, struct seq_file *s, int qid)
 	int inst = 0;
 	struct hal_priv *priv = (struct hal_priv *)dp_port_prop[inst].priv_hal;
 	u32 mode = get_dp_port_info(inst, port_id)->cqe_lu_mode;
+	int match;
 
 	left_n = 1 << (LOOKUP_FIELD_BITS - 4);	/*maximum lookup entried */
 	lookup_match_num = 0;
@@ -337,13 +340,13 @@ int find_pattern(int port_id, struct seq_file *s, int qid)
 	for (r = n; r >= 0; r--) {
 		if (left_n <= 0)
 			break;
-		c_not_care_walkthrought(arr, n, r);
+		match = c_not_care_walkthrought(arr, n, r);
 		DP_DEBUG(DP_DBG_FLAG_LOOKUP, "left_n=%d\n", left_n);
 		if (!left_n)
 			break;
 	}
 
-	for (i = 0; i < lookup_match_num; i++) {
+	for (i = 0; i < match; i++) {
 		if ((qid >= 0) && (qid != lookup_match_qid[i]))
 			continue;
 		if (!f) {
@@ -398,10 +401,10 @@ int lookup_dump31(struct seq_file *s, int pos)
 	return pos;
 }
 
-ssize_t proc_get_qid_via_index31(struct file *file, const char *buf,
+ssize_t proc_get_qid_via_index31(struct file *file, const char __user *buf,
 				 size_t count, loff_t *ppos)
 {
-	int err = 0, len = 0;
+	size_t len;
 	char data[100];
 	unsigned int lookup_index;
 	unsigned int qid = 0;
@@ -412,20 +415,8 @@ ssize_t proc_get_qid_via_index31(struct file *file, const char *buf,
 		return -EPERM;
 	len = (count >= sizeof(data)) ? (sizeof(data) - 1) : count;
 	DP_DEBUG(DP_DBG_FLAG_LOOKUP, "len=%d\n", len);
-
-	if (len <= 0) {
-		err = -EFAULT;
-		pr_err("Wrong len value (%d)\n", len);
-		return count;
-	}
-
-	if (copy_from_user(data, buf, len)) {
-		err = -EFAULT;
-		pr_err("copy_from_user fail");
-		return count;
-	}
-
-	data[len - 1] = 0; /* Make string */
+	len -= copy_from_user(data, buf, len);
+	data[len] = 0; /* Make string */
 	num = dp_split_buffer(data, param_list, ARRAY_SIZE(param_list));
 
 	if (num <= 1)
diff --git a/drivers/net/datapath/umt/umt.c b/drivers/net/datapath/umt/umt.c
index 7cff7ae69659..3afa41afb1f4 100644
--- a/drivers/net/datapath/umt/umt.c
+++ b/drivers/net/datapath/umt/umt.c
@@ -425,11 +425,6 @@ static int umt_set_rxmux(struct umt_priv *priv, struct umt_port *umt_p)
 		} else {
 			umt_reg_update_bit(priv, UMT_MODE, P0_CNT_MODE_OFF, 1);
 			for (i = 0; i < res->dma_ch_num; i++) {
-				if (rxmux > max_mux) {
-					dev_err(priv->dev, "UMT Port %u invalid rxmux: %d\n",
-						id, rxmux);
-					return -EINVAL;
-				}
 				umt_reg_write_mask(priv, UMT_RX0_MUX,
 						   to_mask(UMT_MUX_WIDTH),
 						   umt_rx0_mux_off(priv, i),
